### DISCUSSION TITLE
Having a problem with wavplayer

### DISCUSSION THREAD

**Discussion by bouceka** [micropython/micropython#13059] [ESP32]
Hi ✋ , I'm having problem to play audio with ESP32. I pretty much followed [this repo ](https://github.com/miketeachman/micropython-i2s-examples/blob/master/examples/easy_wav_player.py). However, I am getting this error:
```
Root directory: ()
Traceback (most recent call last):
  File "wavplayer.py", line 61, in i2s_callback
OSError: [Errno 5] EIO
```

Code: 
```python
# The MIT License (MIT)
# Copyright (c) 2022 Mike Teachman
# https://opensource.org/licenses/MIT
#
# Purpose:  Play a WAV audio file out of a speaker or headphones
#

import os
import time
from machine import Pin
import uos

from wavplayer import WavPlayer

if os.uname().machine.count("ESP32"):
#     from machine import SDCard
    from sdcard import SDCard
    from machine import SoftSPI
    # ======= SD CARD CONFIGURATION =======
#     os.umount("/sd")
    spisd = SoftSPI(-1,sck=Pin(18), miso=Pin(19), mosi=Pin(23))
    sd = SDCard(spisd, Pin(5))
    print ('Root directory: ()'.format(os.listdir ()))
    vfs = os.VfsFat(sd)
    os.mount(vfs,"/sd")
    # ======= SD CARD CONFIGURATION =======

    # ======= I2S CONFIGURATION =======
    SCK_PIN = 26
    WS_PIN = 25
    SD_PIN = 22
    I2S_ID = 0
    BUFFER_LENGTH_IN_BYTES = 40000
    # ======= I2S CONFIGURATION =======

else:
    print("Warning: program not tested with this board")

wp = WavPlayer(
    id=I2S_ID,
    sck_pin=Pin(SCK_PIN),
    ws_pin=Pin(WS_PIN),
    sd_pin=Pin(SD_PIN),
    ibuf=BUFFER_LENGTH_IN_BYTES,
)

wp.play("music-16k-16bits-stereo.wav", loop=False)
# wait until the entire WAV file has been played
while wp.isplaying() == True:
    # other actions can be done inside this loop during playback
    pass

wp.play("music-16k-16bits-mono.wav", loop=False)
time.sleep(10)  # play for 10 seconds
wp.pause()
time.sleep(5)  # pause playback for 5 seconds
wp.resume()  # continue playing to the end of the WAV file
```

Using:
1. Micropython v1.21.0 (2023-10-05)
2. Esp32 
3. Amplifier PAM 8403
4. Micro SD Adapter

ESP32 reads the SD Card. I can see it in Thonny. I tried to play music using C and it work, but I cannot figure out how to make it work with Python.

**Comment by peterhinch** [comment]
The PAM 8403 does not appear to be an I2S amplifier: it has analog inputs. You need an I2S DAC between the ESP32 and the PAM 8403.

**Reply by bouceka** [reply]
Hi Peter, my bad, the amplifier I'm using is this:
<img width="233" alt="Screenshot 2023-11-24 at 10 25 36 AM" src="https://github.com/micropython/micropython/assets/49728880/eff7820d-6a67-4f89-b1ab-5e085942dbd7">

The convertor ES7148 should be able to handle I2S.

**Comment by steveh127** [comment]
A couple of suggestions.

Double check your wiring, not only that connections are correct but also that they are as short as possible and sound. I2S is moving a lot of data fast and is quite sensitive to poor wiring.

Looking at the ES7148 datasheet  it also needs a system clock input as well as the standard I2S connections. This is usually labelled MCLK. Your module may be providing this but I doubt it. If you are using an ESP32 you'll need to generate this yourself. There's a recent discussion on this, #11675, that should point you in the right direction.

**Reply by peterhinch** [reply]
The comment about wiring is true - in spades. I found standard jumper leads useless and designed a PCB. I think the clock signal drives a phase locked loop on the DAC chip to produce a higher frequency clock. This causes the DAC to be very critical of the edges on the clock.
